### PR TITLE
Migrate Opportunities api to Nest

### PR DESCRIPTION
**Pull request by AlreadyBored**
**🟢 Add `deploy` label if you want to deploy this Pull Request to staging environment**

#### 🧑‍⚖️ Pull Request Naming Convention

- Title should follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- Do not put issue id in title
- Do not put WIP in title. Use [Draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/) functionality
- Consider to add `area:*` label(s)

* [x] I followed naming convention rules

---

#### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Performance optimization
- [x] Refactoring
- [ ] Test Case
- [ ] Documentation update
- [ ] Other

#### 🔗 Related issue link

https://github.com/rolling-scopes/rsschool-app/issues/1754

#### 💡 Background and solution

Removed opportunities-related code from `server` and migrated to `nestjs` following endpoints:  
- save resume (POST `/opportunities/{githubId}/resume`)
- update resume validity status (POST `/opportunities/status`)
- update resume visibility status (POST `/opportunities/visibility`)

#### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Database migration is added or not needed
- [x] Documentation is updated/provided or not needed
- [x] Changes are tested locally

### CODE REVIEW

File: nestjs/src/opportunities/opportunities.controller.ts [hunk 1]
```diff
@@ -22,7 +24,11 @@ import { CurrentRequest, DefaultGuard, RequiredRoles, Role, RoleGuard } from 'sr
 import { ApplicantResumeDto } from './dto/applicant-resume.dto';
 import { ConsentDto } from './dto/consent.dto';
 import { ResumeDto } from './dto/resume.dto';
+import { StatusDto } from './dto/status.dto';
+import { VisibilityDto } from './dto/visibility.dto';
+//import { FormDataDto } from './dto/form-data.dto';
```
Comment (valerydluski) [inline nestjs/src/opportunities/opportunities.controller.ts:29]: is it necessary or left after the migration?

Comment (AlreadyBored) [reply nestjs/src/opportunities/opportunities.controller.ts:29]: nope, this should be deleted, thx

File: nestjs/src/opportunities/dto/status.dto.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { ApiProperty } from '@nestjs/swagger';
+
+export class StatusDto {
+  constructor(expires: number) {
+    this.data = expires;
+  }
+  @ApiProperty()
+  public data: number;
```
Comment (aaliakseyenka) [inline nestjs/src/opportunities/dto/status.dto.ts:8]: consider naming `expires` or whatever it should mean

File: nestjs/package.json [hunk 3]
```diff
@@ -47,6 +47,7 @@
     "ical-generator": "3.5.1",
     "jsonwebtoken": "8.5.1",
     "lodash": "4.17.21",
+    "luxon": "3.0.3",
```
Comment (aaliakseyenka) [inline nestjs/package.json:50]: we already use `dayjs `

File: client/src/modules/Opportunities/components/EditCv/index.tsx [hunk 4]
```diff
@@ -47,9 +46,9 @@ export const EditCV = (props: Props) => {
   const visibleCoursesFormRef: RefObject<FormInstance> = createRef();
 
   const submitData = async (data: AllUserCVData) => {
-    const newData = await cvService.saveResumeData(data);
+    const newData = await service.saveResume(props.githubId, { data });
 
-    const { userData, contacts, visibleCourses } = splitDataForForms(newData);
+    const { userData, contacts, visibleCourses } = splitDataForForms(newData.data as AllUserCVData);
```
Comment (aaliakseyenka) [inline client/src/modules/Opportunities/components/EditCv/index.tsx:51]: curious why do u need casting: shouldn't it be same type?

File: nestjs/src/opportunities/dto/form-data.dto.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  @ApiProperty({ nullable: true, type: String })
+  public selfIntroLink: string | null;
+  @ApiProperty({ nullable: true, type: String })
+  public skype: string | null;
+  @ApiProperty({ nullable: true, type: String })
+  public startFrom: string | null;
+  @ApiProperty({ nullable: true, type: String })
+  public telegram: string | null;
+  @ApiProperty({ nullable: true, type: String })
+  public website: string | null;
```
Comment (aaliakseyenka) [inline nestjs/src/opportunities/dto/form-data.dto.ts:62]: consider adding line breaks

File: nestjs/src/opportunities/opportunities.controller.ts [hunk 6]
```diff
@@ -79,6 +105,24 @@ export class OpportunitiesController {
     return new ConsentDto(data);
   }
 
+  @Post('/status')
+  @ApiOperation({ operationId: 'updateStatus' })
+  @ApiOkResponse({ type: StatusDto })
+  @UseGuards(DefaultGuard)
+  public async updateStatus(@Req() req: CurrentRequest) {
```
Comment (aaliakseyenka) [inline nestjs/src/opportunities/opportunities.controller.ts:112]: IMO looks not good: `update status`, but inside we do some implicit prolongation. I am not aware about the `opportunity` feature, but probably smth  like `prolong` would be better

File: nestjs/src/opportunities/opportunities.controller.ts [hunk 7]
```diff
@@ -49,6 +54,27 @@ export class OpportunitiesController {
...
   }
 
+  @Patch('/:githubId/resume')
+  @ApiOperation({ operationId: 'saveResume' })
+  @ApiForbiddenResponse()
+  @ApiBadRequestResponse()
+  @ApiNotFoundResponse()
+  @ApiOkResponse({ type: Resume })
+  @UseGuards(DefaultGuard)
+  // TODO: deal with 500 status code validation issue on assigning dto FormDataDto type + @Body() decorator
```
Comment (aaliakseyenka) [inline nestjs/src/opportunities/opportunities.controller.ts:64]: you dto has constructor, if i recall correctly nest won't pass your data into it

File: client/src/modules/Opportunities/components/EditCv/index.tsx (hunk 4 above)
Comment (AlreadyBored) [reply client/src/modules/Opportunities/components/EditCv/index.tsx:51]: this is due to incompatibility of nullable enums - I tried various approaches but it doesn't set enum | null in client api

File: nestjs/src/opportunities/opportunities.controller.ts (hunk 7 above)
Comment (AlreadyBored) [reply nestjs/src/opportunities/opportunities.controller.ts:64]: I tried to use dto without constructor, it has different errors
whitelist-allowed-error
whitelist-error

Comment (apalchys) [reply nestjs/src/opportunities/opportunities.controller.ts:64]: @AlreadyBored you need to define validation for your fields. You just defined OpenAPI decorators

Comment (apalchys) [reply nestjs/src/opportunities/opportunities.controller.ts:64]: actually, in your case I think the problem is constructor. You need to split request and response dtos.

Comment (AlreadyBored) [reply nestjs/src/opportunities/opportunities.controller.ts:65]: I removed constructor and it still doesn't work, please take a look at latest version :(

Comment (AlreadyBored) [reply nestjs/src/opportunities/opportunities.controller.ts:65]: I tried to remove constructor, it still doesn't work, please take a look at the latest version. We have a lot of similar dtos without validation decorators and it works well Oo

Comment (AlreadyBored) [reply nestjs/src/opportunities/opportunities.controller.ts]: Resolved

**Review by apalchys — APPROVED** [review]

**Review by aaliakseyenka — APPROVED** [review]

**Review by valerydluski — APPROVED** [review]

**Pull request merged**